syntax = "proto3";

import "data/common.proto";

package data;

option cc_enable_arenas = true;
option java_outer_classname = "ProcessSummaryProtos";

message ProcessSummary {

    enum ProcessOp {
        UNKNOWN   = 0;
        CLONE     = 1;
        EXECVE    = 2;
        EXIT      = 3;
        KILL      = 4;
        SETNS     = 5;
        SETID     = 6;
        MODULE    = 7;
        reserved    8; // Formerly SESSION.
        SOCKET    = 9;
        CAP       = 10;
        PERM      = 11;
    }

    enum SocketFamily {
        UNKNOWN_FAMILY = 0;
        IP   = 1;
        UNIX = 2;
    }

    message ProcessOperation {
        message Details {
            message Socket {
                SocketFamily family = 1;
                uint32 client_port = 2;
                uint32 server_port = 3;
            }
            message Module {
                string name = 1;
            }
            message PermissionChange {
                string path = 1;
            }
            Socket socket = 1; // operation == SOCKET (optional)
            Module module = 2; // operation == MODULE (optional)
            PermissionChange permissionChange = 3; // operation == PERM (optional)
        }

        ProcessOp operation = 1; // based off of syscall
        string syscall      = 2; // evt_type
        string syscall_args = 3; // evt_type
        bool failure        = 4; // evt_res
        Details details     = 5;
    }

    message Container {
        string id       = 1; // container_id
        bool privileged = 2; // container_privileged
        UUID cluster_id = 3;
    }

    UUID id              = 1;
    Container container  = 2;
    Process process      = 3;
    repeated ProcessOperation ops = 4;
    int64 timestamp      = 5;
}
